# Equation Solving

First we will learn how to an equation in t when equated to 0
$$v_0t - \frac{1}{2}gt^2 = 0$$

In [2]:
from sympy import (
    symbols, #Define mathematical symbols for symbolic math
    solve, 
    Rational, #define rational numbers
    lambdify #Turn symbolic expressions to Python functions
)

In [3]:
t, v0, g = symbols('t v0 g')

In [22]:
y = v0*t - Rational(1, 2)*g*t**2

In [23]:
roots = solve(y , t)

In [24]:
roots

[0, 2*v0/g]

We can easily check the answer by inserting the values in the expression. This is done as shown below

In [7]:
y.subs(t, roots[0])

0

In [8]:
y.subs(t, roots[1])

0

# Taylor Polynomial

A taylor polynomial of order `n` for an expression `e` in a variable `t` around the point `t0` is computed by `e.series(t, t0, n)` 

Let's test this on $e^t$ and $e^{sin(t)}$

In [9]:
from sympy import sin, cos, exp
f = exp(t)
f.series(t, 0, 3)

1 + t + t**2/2 + O(t**3)

In [10]:
f = exp(sin(t))
f.series(t, 0, 8)

1 + t + t**2/2 - t**4/8 - t**5/15 - t**6/240 + t**7/90 + O(t**8)

Output of the mathematical expression in LATEX typesetting is also possible

In [12]:
from sympy import latex
print latex(f.series(t, 0, 7))

1 + t + \frac{t^{2}}{2} - \frac{t^{4}}{8} - \frac{t^{5}}{15} - \frac{t^{6}}{240} + \mathcal{O}\left(t^{7}\right)


Using the above output, I will create this expression 
$$1 + t + \frac{t^{2}}{2} - \frac{t^{4}}{8} - \frac{t^{5}}{15} - \frac{t^{6}}{240} + \mathcal{O}\left(t^{7}\right)$$

Lastly, we should also see tools for expanding and simplifying

In [13]:
from sympy import simplify, expand
x, y = symbols('x y')
f = -sin(x)*sin(y) + cos(x)*cos(y)
simplify(f)

cos(x + y)

In [14]:
expand(cos(x+y)) #Without any trignometric hint

cos(x + y)

In [16]:
expand(cos(x+y), trig=True) # With some trignometric hint

-sin(x)*sin(y) + cos(x)*cos(y)

In [18]:
a,b = symbols('a, b')
f = (a+b)**2
expand(f)

a**2 + 2*a*b + b**2